kernel/meta/gcc.cpp

```diff
@@ -0,0 +1,56 @@
+/**
+ * @file cxx.cpp
+ * @author Keeton Feavel (keetonfeavel@cedarville.edu)
+ * @brief local static variable guards
+ * @version 0.1
+ * @date 2021-07-24
+ * 
+ * @copyright Copyright the Panix Contributors (c) 2021
+ * 
+ * References:
+ *     https://wiki.osdev.org/C++#Local_Static_Variables_.28GCC_Only.29
+ *
+ */
+#include <stddef.h>
+#include <stdint.h>
+#include <dev/serial/rs232.hpp>
+
+// Function prototypes
+extern "C" void __cxa_pure_virtual();
+
+// These static variable guard functions are only used by
+// GCC (according to the Wiki). If Panix ever switches to
+// or supports Clang, we want to be prepared.
+#ifdef __GNUG__
+namespace __cxxabiv1 {
+
+// The ABI requires a 64-bit type.
+__extension__ typedef int64_t __guard __attribute__((mode(__DI__)));
+
+// Function prototypes
+extern "C" int __cxa_guard_acquire(__guard* g);
+extern "C" void __cxa_guard_release(__guard* g);
+extern "C" void __cxa_guard_abort(__guard* g);
+
+extern "C" int __cxa_guard_acquire(__guard* g)
+{
+    return __atomic_test_and_set(g, __ATOMIC_RELEASE);
+}
+
+extern "C" void __cxa_guard_release(__guard* g)
+{
+    __atomic_clear(g, __ATOMIC_RELEASE);
+}
+
+extern "C" void __cxa_guard_abort(__guard* g)
+{
+    __atomic_clear(g, __ATOMIC_RELEASE);
+}
+
+} // !__cxxabiv1
+#endif
+
+extern "C" void __cxa_pure_virtual()
+{
+    rs232::printf("[%s] Pure virtual called!\n", __FUNCTION__);
```

Comment:
We *must* panic here, as this is undefined behavior. Also, `__FUNCTION__` will always be `__cxa_pure_virtual`, so I'm not sure how useful that is. The panic screen's back trace will reveal who called it.

Reply:
Okay, I was wondering. Originally I had a panic there, but I couldn't remember if it could happen normally. Good to know.

Reply:
We do need to figure out how to get symbols working in the panic backtrace.

Reply:
Yeah, probably should open an issue for that.